File: back-end/src/main/java/es/ull/project/domain/enumerate/FacilityStatus.java
package es.ull.project.domain.enumerate;


import java.util.Random;

/**
 * FacilityStatus
 *
 * Represents the current lifecycle state of a facility.
 * It defines whether a facility can be used to assign containers.
 * CANDIDATE, PLANNED, OPEN, or DISCARDED.
 */
public enum FacilityStatus {

    /**
     * Facility is a candidate and not yet selected.
     */
    CANDIDATE,

    /**
     * Facility is planned but not yet open.
     */
    PLANNED,

    /**
     * Facility is open and operational.
     */
    OPEN,

    /**
     * Facility has been discarded and cannot be used.
     */
    DISCARDED;

    /**
     * Returns the FacilityStatus that matches the given string.
     *
     * @param stringToCheck String representation of the status.
     * @return FacilityStatus matching the string.
     * @throws IllegalArgumentException if the string is null or invalid.
     */
    public static FacilityStatus fromString(String stringToCheck) {
        if (stringToCheck == null) {
            throw new IllegalArgumentException();
        }
        stringToCheck = stringToCheck.trim().toUpperCase();
        for (FacilityStatus status : values()) {
            if (status.name().equals(stringToCheck)) {
                return status;
            }
        }
        throw new IllegalArgumentException();
    }

    /**
     * Returns the ordinal index of the given status string.
     *
     * @param stringToCheck String representation of the status.
     * @return Ordinal index of the status.
     */
    public static int indexOf(String stringToCheck) {
        return FacilityStatus
                .fromString(stringToCheck)
                .ordinal();
    }

    /**
     * Checks whether the given string is a valid FacilityStatus.
     *
     * @param stringToCheck String to validate.
     * @return True if valid, false otherwise.
     */
    public static boolean isValid(String stringToCheck) {
        if (stringToCheck == null) {
            return false;
        }
        stringToCheck = stringToCheck.trim().toUpperCase();
        for (FacilityStatus status : values()) {
            if (status.name().equals(stringToCheck)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Returns a random FacilityStatus.
     * Useful for testing purposes.
     *
     * @return Random FacilityStatus.
     */
    public static FacilityStatus random() {
        return values()[new Random().nextInt(values().length)];
    }

    /**
     * Checks whether this status allows service assignments.
     *
     * @return True if containers can be assigned to the facility.
     */
    public boolean allowsServiceAssignments() {
        return this == PLANNED || this == OPEN;
    }

    /**
     * Checks whether this status is DISCARDED.
     *
     * @return True if the facility is discarded.
     */
    public boolean isDiscarded() {
        return this == DISCARDED;
    }
}

File: front-end/src/domain/enumerate/facility-status.ts
/**
 * FacilityStatus
 *
 * Represents the current lifecycle state of a facility.
 *
 * Values:
 * - CANDIDATE: Facility is a candidate and not yet selected.
 * - PLANNED: Facility is planned but not yet open.
 * - OPEN: Facility is open and operational.
 * - DISCARDED: Facility has been discarded and cannot be used.
 */
export enum FacilityStatus {
  /** Facility is a candidate and not yet selected. */
  CANDIDATE = 'CANDIDATE',
  /** Facility is planned but not yet open. */
  PLANNED = 'PLANNED',
  /** Facility is open and operational. */
  OPEN = 'OPEN',
  /** Facility has been discarded and cannot be used. */
  DISCARDED = 'DISCARDED',
}

/**
 * Returns an array with all FacilityStatus values.
 * @returns {FacilityStatus[]} array with all allowed facility statuses.
 */
export function facilityStatusValues(): FacilityStatus[] {
  return Object.values(FacilityStatus) as FacilityStatus[];
}

/**
 * Parse a string into FacilityStatus.
 * @param {string} [s] String representation of the status.
 * @returns {FacilityStatus} matching facility status.
 * @throws {Error} if the provided string is undefined or invalid.
 */
export function facilityStatusFromString(s?: string): FacilityStatus {
  if (!s) throw new Error('Facility status is not defined');
  const key = s.trim().toUpperCase();
  for (const v of facilityStatusValues()) {
    if (v === key) return v;
  }
  throw new Error('Facility status is invalid');
}

/**
 * Return the ordinal index of a given status string.
 * @param {string} [s] String representation of the status.
 * @returns {number} ordinal index of the status.
 */
export function facilityStatusIndexOf(s?: string): number {
  return facilityStatusValues().indexOf(facilityStatusFromString(s));
}

/**
 * Check whether a string is a valid FacilityStatus.
 * @param {string} [s] String to validate.
 * @returns {boolean} True if valid, false otherwise.
 */
export function isFacilityStatus(s?: string): boolean {
  if (!s) return false;
  const key = s.trim().toUpperCase();
  for (const v of facilityStatusValues()) {
    if (v === key) return true;
  }
  return false;
}

/**
 * Return a random FacilityStatus (useful for testing).
 * @returns {FacilityStatus} random facility status.
 */
export function facilityStatusRandom(): FacilityStatus {
  const values = facilityStatusValues();
  return values[Math.floor(Math.random() * values.length)]!;
}

/**
 * Checks whether the given status allows service assignments.
 * @param {FacilityStatus} status Status to check.
 * @returns {boolean} True if containers can be assigned to the facility.
 */
export function facilityStatusAllowsServiceAssignments(status: FacilityStatus): boolean {
  return status === FacilityStatus.PLANNED || status === FacilityStatus.OPEN;
}

/**
 * Checks whether the given status is DISCARDED.
 * @param {FacilityStatus} status Status to check.
 * @returns {boolean} True if the facility is discarded.
 */
export function facilityStatusIsDiscarded(status: FacilityStatus): boolean {
  return status === FacilityStatus.DISCARDED;
}

/**
 * Returns the UI color for a given facility status.
 * @param {FacilityStatus} status - Facility status.
 * @returns {string} Color name suitable for Vuetify components.
 */
export function facilityStatusColor(status: FacilityStatus | string): string {
  const s = typeof status === 'string' ? facilityStatusFromString(status) : status;
  switch (s) {
    case FacilityStatus.CANDIDATE: return 'grey';
    case FacilityStatus.PLANNED: return 'orange';
    case FacilityStatus.OPEN: return 'green';
    case FacilityStatus.DISCARDED: return 'red';
    default: return 'grey';
  }
}

/**
 * Returns the options for a facility status selector.
 * @param {Function} t Translation function.
 * @returns {{ title: string; value: FacilityStatus }[]} array of options.
 */
export function facilityStatusToOptions(t: (key: string) => string): { title: string; value: FacilityStatus }[] {
  return facilityStatusValues().map((status) => ({
    title: t(`facility.add.statuses.${status}`),
    value: status,
  }));
}
